connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -830,6 +830,66 @@ class Dataset[T] private[sql] (
     builder.setJoinType(proto.Join.JoinType.JOIN_TYPE_CROSS)
   }
 
+  /**
+   * Joins this Dataset returning a `Tuple2` for each pair where `condition` evaluates to
+   * true.
+   *
+   * This is similar to the relation `join` function with one important difference in the
+   * result schema. Since `joinWith` preserves objects present on either side of the join, the
+   * result schema is similarly nested into a tuple under the column names `_1` and `_2`.
+   *
+   * This type of join can be useful both for preserving type-safety with the original object
+   * types as well as working with relational data where either side of the join has column
+   * names in common.
+   *
+   * @param other Right side of the join.
+   * @param condition Join expression.
+   * @param joinType Type of join to perform. Default `inner`. Must be one of:
+   *                 `inner`, `cross`, `outer`, `full`, `fullouter`,`full_outer`, `left`,
+   *                 `leftouter`, `left_outer`, `right`, `rightouter`, `right_outer`.
+   *
+   * @group typedrel
+   * @since 3.5.0
+   */
+  def joinWith[U](other: Dataset[U], condition: Column, joinType: String): Dataset[(T, U)] = {
```

Comment:
shall we add a new logical plan `UnresolvedJoinWith` to simplify the client implementation?

Reply:
That is the thing. That would mean I also have to add this to the protocol. It is okay I guess. However, we already have tried that in a different PR (not the new logical plan), and my main gripe with this is that we are adding a very very scala specific thing to both the protocol and the analyzer, for something that can easily be done on the client with a number of primitives (access by ordinal & explicit references). I could probably live without the latter, if we make ordinal access a bit more flexible (I need access to the first and last column of the df).

Reply:
Then can we add that expression to the protocol to do positional column lookup? I'm a bit worried about making the column resolution more complex with `UnresolvedAttribute`. We already have special expressions `GetColumnByOrdinal` and `GetViewColumnByNameAndOrdinal` to find-tune the column lookup logic and we can add more.